python/lib/core/dmod/core/common/collections/catalog.py

```diff
@@ -0,0 +1,651 @@
+
+    def __ge__(self, other):
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} >= {other.__class__.__name__}")
+
+        return self.last_accessed >= other.last_accessed
+
+    def __gt__(self, other):
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} < {other.__class__.__name__}")
+
+        return self.last_accessed >= other.last_accessed
+
+    def __str__(self):
+        return f"CacheEntry: {self.identifier} @ {self.last_accessed.strftime('%Y-%m-%d %H:%M:%S')}"
+
+    def __repr__(self):
+        return "{" + \
+            f'"identifier": {self.__data}, ' \
+            f'"data_hash": {self.__data_hash}, ' \
+            f'"last_accessed": "{self.last_accessed}"' + \
+            "}"
+
+
+# Part of Issue https://github.com/NOAA-OWP/DMOD/issues/434
+#   "Make metric computations asynchronous by location"
+class InputCatalog(EventfulMap[str, CatalogEntry[HashableTypeT]], typing.Generic[HashableTypeT]):
+    """
+    A base class that implements a caching mechanism that organizes entries based on access time and fires events
+    when adding, removing, accessing, and updating entries
+    """
+
+    def __init__(
+        self,
+        max_size: int = 100,
+        *,
+        values: typing.Mapping[str, typing.Union[HashableTypeT, CatalogEntry[HashableTypeT], None]] = None
+    ):
+        """
+        Constructor
+
+        Args:
+            max_size: The maximum number of items that this cache may contain. Values other than positive numbers will
+                cause the cache to become unbounded
+            values: Preexisting data to add to the cache
+            kwargs: Additional items to store in the cache
+        """
+        if values is None:
+            values = {}
+
+        super().__init__(contents=values)
+
+        self.__earliest: typing.Optional[CatalogEntry[HashableTypeT]] = None
+        self.__max_size = max_size if isinstance(max_size, (int, float)) and max_size > 0 else math.inf
+
+    async def commit(self):
+        """
+        Complete all asynchronous tasks that have yet to be awaited
+        """
+        await super().commit()
+        self.__update_earliest()
+
+    def touch(self, identifier: str, new_time: datetime = None):
+        """
+        Update cache entry access time
+
+        Updating the access time will prevent the entry from being removed upon addition if items need to be removed.
+
+        Triggered asynchronous tasks will be stored for later completion
+
+        Args:
+            identifier: The identifier for the entry to update
+            new_time: The time to set the access time to. Defaults to now
+        """
+        try:
+            self.lock()
+
+            if identifier in self:
+                self.inner_map()[identifier].touch(new_time)
+            else:
+                self[identifier] = None
```

Comment:
I think this is a dead lock condition. You already have the lock at this point and in `__setitem__` you try to acquire the lock again.